### PR TITLE
[Integration][Kubecost][Opencost] Add support for filters

### PR DESCRIPTION
**Pull request by lordsarcastic**
# Description

[Docs PR](https://github.com/port-labs/port-docs/pull/1347)

**What:**
- Added `filter` params to Opencost `cloudcost` and Kubecost `cloud` and `kubesystem` kind.
- Implemented separate config for v1 and v2 selectors for Kubecost `cloud` and `kubesystem` kind

**Why:**
- Opencost and Kubecost added the `filter` feature which is a much more powerful aggregation parameter as opposed to the existing fields, adding this field allows users to be much more flexible with what data they ingest into Port
- The Kubecost integration, while currently on v1 as set:
```
KUBECOST_API_VERSION_1 = "v1"
```
Uses the same selector to handle filters for both `kubesystem` and `cloud` kinds which should be separate. In addition, the [params accepted by v1](https://docs.kubecost.com/v/1.0x/apis/apis-overview/api-allocation) is not the same as [that on v2](https://docs.kubecost.com/apis/monitoring-apis/api-allocation); this applies to both kinds.

**How:**
To address these needs, different selectors are defined for each kind and for each version and can be used when needed

## Type of change

Please leave one option from the following and delete the rest:

- [x] New feature (non-breaking change which adds functionality)
- [x] Non-breaking change (fix of existing functionality that will not change current behavior)

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

- [Kubecost v1 Allocation API](https://docs.kubecost.com/v/1.0x/apis/apis-overview/api-allocation)
- [Kubecost v2 Allocation API](https://docs.kubecost.com/apis/monitoring-apis/api-allocation)
- [Kubecost v1 Cloudcost API](https://docs.kubecost.com/v/1.0x)
- [Kubecost v2 Cloudcost API](https://docs.kubecost.com/apis/monitoring-apis/cloud-cost-api)
- [Opencost Cloudcost API](https://www.opencost.io/docs/integrations/api#cloudcost)

**Note:** Kubecost docs have a slight error in their documentation, I have [raised an issue](https://github.com/kubecost/docs/issues/1088) on their repository for that.

**Mapping config example for v1 `kubesystem` kind**:

```yaml showLineNumbers
createMissingRelatedEntities: true
deleteDependentEntities: true
resources:
  - kind: kubesystem
    selector:
      query: "true"
      window: "7d"
      aggregate: service
      step: "1d" # this will result in 7 entries which is 1 day over the course of `window` time period: 7 days
      accumulate: false
      idle: true
      external: false
      filterLabels: "app:internal-service"
      filterServices: "notification,account,functions"
    port:
      entity:
        mappings:
          blueprint: '"kubecostResourceAllocation"'
          identifier: .name
          title: .name
          properties:
            cluster: .properties.cluster
            namespace: .properties.namespace
            startDate: .start
            endDate: .end
            cpuCoreHours: .cpuCoreHours
            cpuCost: .cpuCost
            cpuEfficiency: .cpuEfficiency
            gpuHours: .gpuHours
            gpuCost: .gpuCost
            networkCost: .networkCost
            loadBalancerCost: .loadBalancerCost
            pvCost: .pvCost
            pvBytes: .pvBytes
            ramBytes: .ramBytes
            ramCost: .ramCost
            ramEfficiency: .ramEfficiency
            sharedCost: .sharedCost
            externalCost: .externalCost
            totalCost: .totalCost
            totalEfficiency: .totalEfficiency
```

**Mapping config example for v2 `kubesystem` kind**:

```yaml showLineNumbers
createMissingRelatedEntities: true
deleteDependentEntities: true
resources:
  - kind: kubesystem
    selector:
      query: "true"
      window: "7d"
      aggregate: service
      step: "1d" # this will result in 7 entries which is 1 day over the course of `window` time period: 7 days
      accumulate: false
      idle: true
      filter: 'labels:"app:internal-service","app:service-2"+service:"notification","account","functions"' # replaces all `filter*` properties needed in v1
    port:
      entity:
        mappings:
          blueprint: '"kubecostResourceAllocation"'
          identifier: .name
          title: .name
          properties:
            cluster: .properties.cluster
            namespace: .properties.namespace
            startDate: .start
            endDate: .end
            cpuCoreHours: .cpuCoreHours
            cpuCost: .cpuCost
            cpuEfficiency: .cpuEfficiency
            gpuHours: .gpuHours
            gpuCost: .gpuCost
            networkCost: .networkCost
            loadBalancerCost: .loadBalancerCost
            pvCost: .pvCost
            pvBytes: .pvBytes
            ramBytes: .ramBytes
            ramCost: .ramCost
            ramEfficiency: .ramEfficiency
            sharedCost: .sharedCost
            externalCost: .externalCost
            totalCost: .totalCost
            totalEfficiency: .totalEfficiency
```

**Mapping config example for v1 `kubecostCloudAllocation` kind**

```yaml showLineNumbers
createMissingRelatedEntities: true
deleteDependentEntities: true
resources:
  - kind: cloud
    selector:
      query: "true"
      window: 7d
      aggregate: service
      filterAccountIDs: "984573291056,374859201234"
      filterLabel: label-one,label-two
      filterServices: notification,account,chat
    port:
      entity:
        mappings:
          blueprint: '"kubecostCloudAllocation"'
          identifier: .properties.provider + "/" + .properties.providerID + "/" + .properties.category + "/" + .properties.service | gsub("[^A-Za-z0-9@_.:\\\\/=-]"; "-")
          title: .properties.provider + "/" + .properties.service
          properties:
            provider: .properties.provider
            accountID: .properties.accountID
            invoiceEntityID: .properties.invoiceEntityID
            startDate: .window.start
            endDate: .window.end
            listCost: .listCost.cost
            listCostPercent: .listCost.kubernetesPercent
            netCost: .netCost.cost
            netCostPercent: .netCost.kubernetesPercent
            amortizedNetCost: .amortizedNetCost.cost
            amortizedNetCostPercent: .amortizedNetCost.kubernetesPercent
            invoicedCost: .invoicedCost.cost
            invoicedCostPercent: .invoicedCost.kubernetesPercent

```

**Mapping config example for v2 `kubecostCloudAllocation` kind**

```yaml showLineNumbers
createMissingRelatedEntities: true
deleteDependentEntities: true
resources:
  - kind: cloud
    selector:
      query: "true"
      window: 7d
      aggregate: service
      accumulate: false
      filter: 'labels:"app:internal-service","app:service-2"+service:"notification","account","functions"' # replaces all `filter*` properties needed in v1
    port:
      entity:
        mappings:
          blueprint: '"kubecostCloudAllocation"'
          identifier: .properties.provider + "/" + .properties.providerID + "/" + .properties.category + "/" + .properties.service | gsub("[^A-Za-z0-9@_.:\\\\/=-]"; "-")
          title: .properties.provider + "/" + .properties.service
          properties:
            provider: .properties.provider
            accountID: .properties.accountID
            invoiceEntityID: .properties.invoiceEntityID
            startDate: .window.start
            endDate: .window.end
            listCost: .listCost.cost
            listCostPercent: .listCost.kubernetesPercent
            netCost: .netCost.cost
            netCostPercent: .netCost.kubernetesPercent
            amortizedNetCost: .amortizedNetCost.cost
            amortizedNetCostPercent: .amortizedNetCost.kubernetesPercent
            invoicedCost: .invoicedCost.cost
            invoicedCostPercent: .invoicedCost.kubernetesPercent

```

### CODE REVIEW

File: integrations/opencost/integration.py [hunk 1]
```diff
@@ -134,6 +134,13 @@ class CloudCostSelector(Selector):
                 description="Step size of the accumulation.",
             )
         )
+        filter: str | None = Field(
+            description=(
+                "Filter results by any category which that can be aggregated by,"
+                " can support multiple filterable items in the same category in"
+                " a comma-separated list."
+            ),
```
Comment (Tankilevitch) [inline integrations/opencost/integration.py:142]: I think we should support list object and parse it to string

File: integrations/kubecost/integration.py [hunk 2]
```diff
@@ -56,125 +55,309 @@ def validate(cls, value: str) -> None:
...
+        filter_providers: str | None = Field(
+            alias="filterProviders", description="Filter for cloud service provider"
+        )
+        filter_label: str | None = Field(
+            alias="filterLabel",
+            description="Filter for a specific label. Does not support filtering for multiple labels at once.",
+        )
+        filter_services: str | None = Field(
+            alias="filterServices",
+            description="Comma-separated list of services to match; e.g. frontend-one,frontend-two will return results with either of those two services",
```
Comment (Tankilevitch) [inline integrations/kubecost/integration.py:93]: should be list in my opinion and parse it to string later on

File: integrations/kubecost/integration.py [hunk 3]
```diff
@@ -56,125 +55,309 @@ def validate(cls, value: str) -> None:
...
+                "week",
+                "month",
+                "yesterday",
+                "lastweek",
+                "lastmonth",
+                "30m",
+                "12h",
+                "7d",
+            ]
+            | DatePairField
```
Comment (Tankilevitch) [inline integrations/kubecost/integration.py:114]: lets move the "selector" class definition outside of the class of resourceConfig and only reference it, see how we do in datadog, azure etc..

File: integrations/kubecost/integration.py [hunk 4]
```diff
@@ -56,125 +55,309 @@ def validate(cls, value: str) -> None:
...
+        )
+        limit: int | None = Field(
+            description="Maximum number of items to return in the result set.",
+        )
+        filter: str | None = Field(
+            description=(
+                "Filter results by any category which that can be aggregated by,"
+                " can support multiple filterable items in the same category in"
+                " a comma-separated list."
+            ),
```
Comment (Tankilevitch) [inline integrations/kubecost/integration.py:135]: same here

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/opencost/integration.py (hunk 1 above)
Comment (MatanHeledPort) [reply integrations/opencost/integration.py:142]: We want to support this format:
```
filter:
  service: AWSSecretsManager,AWSS3,ECS
  provider: AWS
```

File: integrations/kubecost/integration.py [hunk 5]
```diff
@@ -56,125 +55,309 @@ def validate(cls, value: str) -> None:
...
+        aggregate: AggregationField | None = Field(
+            description="Field by which to aggregate the results.",
+        )
+        filter_invoice_entity_ids: str | None = Field(
+            alias="filterInvoiceEntityIDs", description="Filter for account"
+        )
+        filter_account_ids: str | None = Field(
+            alias="filterAccountIDs", description="GCP only, filter for projectID"
+        )
+        filter_providers: str | None = Field(
```
[the remark this answers is not in the document]
Comment (lordsarcastic) [reply integrations/kubecost/integration.py:84]: It's from their end. I've raised an issue on their repository

File: integrations/kubecost/client.py [hunk 6]
```diff
@@ -39,14 +43,9 @@ async def get_kubesystem_cost_allocation(self) -> list[dict[str, Any]]:
...
-            selector = typing.cast(
-                KubecostV1ResourceConfig, event.resource_config
-            ).selector
-        else:
-            selector = typing.cast(
-                KubecostV2ResourceConfig, event.resource_config
-            ).selector
+        selector = typing.cast(
+            KubecostV1ResourceConfig | KubecostV2ResourceConfig, event.resource_config
+        ).selector
```
Comment (Tankilevitch) [inline integrations/kubecost/client.py:48]: this shouldn't be part of the client, but rather on the resync side

File: integrations/kubecost/client.py [hunk 7]
```diff
@@ -73,16 +72,14 @@ async def get_cloud_cost_allocation(self) -> list[dict[str, Any]]:
...
-            ).selector
             url = f"{self.kubecost_host}/model/cloudCost/aggregate"
-        else:
-            selector = typing.cast(
-                CloudCostV2ResourceConfig, event.resource_config
-            ).selector
-            url = f"{self.kubecost_host}/model/cloudCost"
+
+        selector = typing.cast(
+            CloudCostV1ResourceConfig | CloudCostV2ResourceConfig, event.resource_config
```
Comment (Tankilevitch) [inline integrations/kubecost/client.py:81]: same

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/kubecost/client.py (hunk 6 above)
Comment (lordsarcastic) [reply integrations/kubecost/client.py:48]: How do I do that?

Comment (Tankilevitch) [reply integrations/kubecost/client.py:48]: you can create a api_utils.py that will have a method called `transform_kubecost_resource_mapping_to_api_params`

Comment (lordsarcastic) [reply integrations/kubecost/client.py:48]: So, you can't pass the types as an argument, the linter will complain. I put the selector code in the resync side. I'll push so you can see it

File: integrations/kubecost/CHANGELOG.md [hunk 8]
```diff
@@ -7,6 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
+# Port_Ocean 0.1.56 (2024-07-09)
+
+### Features
+
+- Added support for filter in 'cloud' and 'kubesystem' kind (#56)
```
Comment (Tankilevitch) [inline integrations/kubecost/CHANGELOG.md:14]: link to the PR id

File: integrations/kubecost/client.py (hunk 6 above)
Comment (Tankilevitch) [reply integrations/kubecost/client.py]: not sure I understand what you've tried performing that it failed on linting

Comment (Tankilevitch) [reply integrations/kubecost/client.py]: can you show an example?

File: integrations/kubecost/integration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply integrations/kubecost/integration.py]: can you link it?

**Review by Tankilevitch — APPROVED** [review]

**Pull request merged**